Add `bonds` for structure type entries (#465)
Materials-Consortia/OPTIMADE: optimade.rst

    These two sites are correlated (either site 2 or 3 is present).
    However, the presence or absence of sites 0 and 1 is not correlated with the presence or absence of sites 2 and 3 (in the specific example, the pair of sites (0, 2) can occur with 0.2\*0.3 = 6 % probability; the pair (0, 3) with 0.2\*0.7 = 14 % probability; the pair (1, 2) with 0.8\*0.3 = 24 % probability; and the pair (1, 3) with 0.8\*0.7 = 56 % probability).

bonds
~~~~~

- **Description**: A list describing the chemical connectivity in the structure.
- **Type**: list of dictionary with keys:

  - :property:`sites`: a list of integers (REQUIRED)

- **Requirements/Conventions**:

  - **Support**: OPTIONAL support in implementations, i.e., MAY be :val:`null`.
  - **Query**: Support for queries on this property is OPTIONAL.
    If supported, filters MAY support only a subset of comparison operators.

>   - The property SHOULD be :val:`null` for structures for which the bond perception has not been carried out by the implementation.

Author: I have attempted to address this in 1478145 by reformulating to a more generic expression.

>   - :property:`sites` inside each dictionary contains two integers refering to different sites which are deemed chemically connected.

Reviewer: ```suggestion
  - If present, it MUST be a list of dictionaries, each of which represents a chemical bond and MUST have the following key:
    - *sites*: 0-based indexes of the two sites that form a chemical bond.
```

Reviewer: Copied the phrasing from the `assemblies` definition to have a more uniform style.

- **Examples**:

>   - :val:`[ {"sites": [1, 2]} ]`: a structure with a bond between sites 1 and 2.

Reviewer: Maybe it should also be mandated that the indexes in the array are always sorted, e.g. `[0, 8]` is allowed, but not `[8, 0]`?

Author: I have attempted to address this in a49da9e51677bcc08fe224e5f58b55ab79c30ec2.

Reviewer: Thank you. I guess `non-decreasing` is used instead of `ascending` since there are plans to allow sites to be bonded to symmetrical equivalents of themselves or something similar?

Author: Right, I have just described the means for specifying translations. I have replaced "non-decreasing" with "ordered", but if you feel that the "non-decreasing" is better, I can bring it back.

Reviewer: Non-decreasing would be clearer (added relevant suggestion).

    These two sites are correlated (either site 2 or 3 is present).
    However, the presence or absence of sites 0 and 1 is not correlated with the presence or absence of sites 2 and 3 (in the specific example, the pair of sites (0, 2) can occur with 0.2\*0.3 = 6 % probability; the pair (0, 3) with 0.2\*0.7 = 14 % probability; the pair (1, 2) with 0.8\*0.3 = 24 % probability; and the pair (1, 3) with 0.8\*0.7 = 56 % probability).

bonds
~~~~~

- **Description**: A list describing the chemical connectivity of the structure.
- **Type**: list of dictionary with keys:

  - :property:`sites`: a list of integers (REQUIRED)
  - :property:`translations`: a list of list of integers (OPTIONAL)

- **Requirements/Conventions**:

  - **Support**: OPTIONAL support in implementations, i.e., MAY be :val:`null`.
  - **Query**: Support for queries on this property is OPTIONAL.
    If supported, filters MAY support only a subset of comparison operators.
  - The property SHOULD be :val:`null` for structures for which the chemical connectivity is unknown to the implementation.
  - If present, it MUST be a list of dictionaries, each of which represents a chemical bond and MUST have the following key:

>     - *sites*: an ordered list of 0-based indexes of the two sites that form a chemical bond.

Reviewer: ```suggestion
    - *sites*: a non-decreasing list of 0-based indexes of the two sites that form a chemical bond.
```

  - If translations are needed by at least one of the sites of a bond, the following key SHOULD be used:

    - *translations*: a list of two lists of three integers each, defining translations of the sites.
      Omitting this key means that both translation vectors are :val:`[0, 0, 0]`.

> 

Reviewer 2: ```suggestion
    - *distance*: the distance in Angstrom between the two sites after applying the appropriate translation vectors.
    
```

Might be nice to also allow this as an optional field.

- **Examples**:

  - :val:`[ {"sites": [1, 2]} ]`: a structure with a bond between sites 1 and 2.

>   - :val:`[ {"sites": [1, 1], "translations": [ [0, 0, 0], [0, 0, 1] ]} ]`: a 1D polymer.

Reviewer: This property could potentially be used to describe crystal topology which may involve various types of contact distances, but that seems to be a bit a beast of its own (at least according to work done in the `TOPO_CIF` dictionary [1]). I would probably limit it to chemical bonding for now, especially since currently we purposely are trying to avoid specifying the bond type/order.

[1] https://github.com/COMCIFS/TopoCif/blob/main/dictionary/Topology_0.9.5.dic

    These two sites are correlated (either site 2 or 3 is present).
    However, the presence or absence of sites 0 and 1 is not correlated with the presence or absence of sites 2 and 3 (in the specific example, the pair of sites (0, 2) can occur with 0.2\*0.3 = 6 % probability; the pair (0, 3) with 0.2\*0.7 = 14 % probability; the pair (1, 2) with 0.8\*0.3 = 24 % probability; and the pair (1, 3) with 0.8\*0.7 = 56 % probability).

bonds
~~~~~

- **Description**: A list describing the chemical connectivity of the structure.
- **Type**: list of dictionary with keys:

  - :property:`sites`: a list of integers (REQUIRED)

>   - :property:`translations`: a list of list of integers (OPTIONAL)

Reviewer 3: We should probably specify very clearly that translations must be applied to the coordinates of atoms as they are currently given, and not to, e.g., coordinates reduced to the [0;1) unit cell.

  - :property:`distance`: float (OPTIONAL)

- **Requirements/Conventions**:

  - **Support**: OPTIONAL support in implementations, i.e., MAY be :val:`null`.
  - **Query**: Support for queries on this property is OPTIONAL.
    If supported, filters MAY support only a subset of comparison operators.
  - The property SHOULD be :val:`null` for structures for which the chemical connectivity is unknown to the implementation.
  - If present, it MUST be a list of dictionaries, each of which represents a chemical bond and MUST have the following key:

    - *sites*: a non-decreasing list of 0-based indexes of the two sites that form a chemical bond.

>   - If translations are needed by at least one of the sites of a bond, the following key SHOULD be used:

Author: I assume you are talking about `translations`, as `sites` is mandatory?

>   - If translations are needed by at least one of the sites of a bond, the following key SHOULD be used:

Reviewer 4: Works for me :)

> I intentionally omit the bond types as this might be difficult to agree upon

The type of bond is not specified in most of the topology formats but I think this is very inconvenient. So happy to know this is provisional and the type will be specified in a future.

>   - If translations are needed by at least one of the sites of a bond, the following key SHOULD be used:

Reviewer: @merkys Yes, I actually meant `translations` (fixed).

    - *translations*: a list of two lists of three integers each, defining translations of the sites.
      Omitting this key means that both translation vectors are :val:`[0, 0, 0]`.

>   - OPTIONAL key :property:`distance` MAY be used to provide distance in ångström (Å) between the two sites after applying the appropriate translation vectors.

Reviewer: ```suggestion
  - OPTIONAL key :property:`distance` MAY be used to provide the distance in ångströms (Å) between the two sites after applying the appropriate translation vectors.
```
